plugin/src/main/java/org/opennms/plugins/zabbix/ZabbixAgentCollector.java

```diff
@@ -65,48 +64,28 @@ public ZabbixAgentCollector(ZabbixAgentClientFactory clientFactory) {
                 LOG.debug("Processing template with name: {}", template.getName());
-                for (Item item : template.getItems()) {
-                    String value = client.retrieveData(item.getKey()).get();
-                    addValueToMapper(null, item, value, nodeResourceBuilder);
-                }
-
+                templateFutures.add(processItems(client, null, template.getItems(), nodeResourceBuilder));
                 final ZabbixResourceTypeGenerator resourceTypeGenerator = new ZabbixResourceTypeGenerator();
                 for (DiscoveryRule rule : template.getDiscoveryRules()) {
                     try {
-                        final List<Map<String, Object>> entries = client.discoverData(rule.getKey()).get();
-                        // We have some data, let's create a new resource type
-                        final ResourceType resourceType = resourceTypeGenerator.getResourceTypeFor(rule);
-                        for (Map<String, Object> entry : entries) {
-                            // Create a new resource for the entry
-                            final ImmutableCollectionSetResource.Builder<GenericTypeResource> resourceBuilder = ImmutableCollectionSetResource.newBuilder(GenericTypeResource.class)
-                                    .setResource(ImmutableGenericTypeResource.newBuilder()
-                                            .setNodeResource(nodeResource)
-                                            .setType(resourceType.getName())
-                                            .setInstance(resourceTypeGenerator.getIndex(rule, entry))
-                                            .build());
-
-                            // Process the items in the rule
-                            boolean didAddAttribute = false;
-                            for (Item item : rule.getItemPrototypes()) {
-                                final String effectiveKey = ZabbixMacroSupport.evaluateMacro(item.getKey(), entry);
-                                String value = client.retrieveData(effectiveKey).get();
-                                addValueToMapper(rule, item, value, resourceBuilder);
-                                didAddAttribute = true;
-                            }
-
-                            // Only add the resource if we found 1+ items
-                            if (didAddAttribute) {
-                                collectionSetBuilder.addCollectionSetResource(resourceBuilder.build());
-                            }
-                        }
-                    } catch (ZabbixNotSupportedException e) {
+                        templateFutures.add(client.discoverData(rule.getKey())
+                                .thenAccept(list -> {
+                                    try {
+                                        processEntries(client, rule, list, collectionSetBuilder, nodeResource, resourceTypeGenerator);
+                                    } catch (IOException e) {
```

Comment:
IOException shouldn't be thrown by this - should return a future with exception instead.

Reply:
Good catch.